async_simple/Signal.h

```diff
@@ -0,0 +1,453 @@
+    }
+    // Set the current scope's filter.
+    void setFilter(SignalType filter) noexcept { _node->setFilter(filter); }
+
+    // Get the current scope's filter.
+    // if slot is not connect, always return SignalType::all
+    SignalType getFilter() const noexcept { return _node->getFilter(); }
+
+    // Check whether the filtered cancellation signal has emitted.
+    // if return true, the signal has emitted(but handler may hasn't execute
+    // now) if slot is not connect to signal, always return true;
+    bool hasTriggered(SignalType type) const noexcept {
+        return _node->triggered(
+            static_cast<SignalType>(signal()->state() & type));
+    }
+    bool canceled() const noexcept {
+        return hasTriggered(SignalType::terminate);
+    }
+
+    // The slot holds ownership of the corresponding signal, so the signal's
+    // lifetime is always no shorter than the slot's. To extend the signal's
+    // lifetime, you can call signal()->shared_from_this(), or start a new
+    // coroutine with the signal.
+    // return nullptr when signal isn't connect to slot.
+    Signal* signal() const noexcept { return _signal.get(); }
+
+    friend class Signal;
+
+    // bind a sub signal to this slot. allow triggered signal will transfer to
+    // this subSignal.
+    void addSubSignal(Signal* subSignal) {
+        if (subSignal == nullptr) {
+            _node->_transferHandler = nullptr;
+            return;
+        }
+
+        _node->_transferHandler.store(
+            new Signal::TransferHandler(
+                [subSignal = subSignal->weak_from_this()](SignalType type) {
+                    if (auto signal = subSignal.lock(); signal != nullptr) {
+                        signal->emit(type);
+                    }
+                }),
+            std::memory_order_release);
+    }
+    ~Slot() {
+        delete _node->_transferHandler.exchange(nullptr,
+                                                std::memory_order_release);
+    }
+
+private:
+    template <bool should_emplace>
+    std::atomic<Signal::Handler*>* loadHandler(SignalType type) {
+        uint8_t index = std::countr_zero(static_cast<uint64_t>(type));
+        if (auto iter = _handlerTables.find(index);
+            iter != _handlerTables.end()) {
+            return iter->second;
+        }
+        if constexpr (!should_emplace) {
+            return nullptr;
+        } else {
+            auto handlerManager =
+                _node->_handlerManager.load(std::memory_order_acquire);
+            auto newManager = new Signal::HandlerManager(index, handlerManager);
+            // It's ok here, write only has once thread
+            _node->_handlerManager.store(newManager, std::memory_order_release);
+            _handlerTables.emplace(index, &newManager->_handler);
+            return &newManager->_handler;
+        }
+    }
+
+    // TODO: may be flatmap better?
+    std::unordered_map<uint8_t, std::atomic<Signal::Handler*>*> _handlerTables;
```

Comment:
Why atomic? We should have banned manipulating the same slot simultaneously.

Reply:
signal->emit() may modify handler when slot emplace()/clear() handler concurrently, so it should be atomic.